src/test/ui/rfc-2632-const-trait-impl/call-generic-method-nonconst.rs

```diff
@@ -0,0 +1,26 @@
+// FIXME(jschievink): this is not rejected correctly (only when the non-const impl is actually used)
```

Comment:
This is interesting... I don't see how this could be caught by the type system without the test as written also failing, so I'm assuming it's post monomorphization, during actual const eval?

Reply:
Hmm... Shouldn't we have to prove constness of whatever `ImplSource` ends up being used to satisfy a `const` trait bound when calling the `const fn` from a const context?

Of course, *any* `ImplSource` is fine when using it from a non-const context, but when using it from a `const` initializer like here I'd expect only `impl const` to be valid.

Reply:
Sorry. I realize how useless my previous post was.

I'm in full agreement that this should be caught. What I was trying to ask is whether what you encountered was a const eval post-monomorphization error. If it is *not* that, then I'm very confused.

We should have a proper check, and I think the only way to do that is to do it via well-formedness checks, though I'm not sure if we need a new one. Basically we need to add that additional wf-check only from within const contexts.

The alternative (which you are suggesting I think?) is to encode in the existing wf-checks (that check whether a trait bound exists) that we are looking for a `const Trait` impl and not just a `Trait` impl when coming from a const context.

Either way would be fine with me. I think your alternative can work, though I worry that it would be very sublte. On the other hand, it would be the version with the least code duplication, so it's definitely preferrable if we can make it work and leave enough comments for future devs that touch the relevant parts.

Reply:
> I'm in full agreement that this should be caught. What I was trying to ask is whether what you encountered was a const eval post-monomorphization error. If it is not that, then I'm very confused.

Oh, yeah, you get this error when replacing the body of `equals_self` with `*t == *t`:

```
error: any use of this value will cause an error
  --> /home/jonas/dev/rust/src/test/ui/rfc-2632-const-trait-impl/call-generic-method-nonconst.rs:16:5
   |
LL |     *t == *t
   |     ^^^^^^^^
   |     |
   |     calling non-const function `<S as PartialEq>::eq`
   |     inside `equals_self::<S>` at /home/jonas/dev/rust/src/test/ui/rfc-2632-const-trait-impl/call-generic-method-nonconst.rs:16:5
   |     inside `EQ` at /home/jonas/dev/rust/src/test/ui/rfc-2632-const-trait-impl/call-generic-method-nonconst.rs:22:22
...
LL | pub const EQ: bool = equals_self(&S);
   | -------------------------------------
   |
   = note: `#[deny(const_err)]` on by default
```

Reply:
Wonderful!

So... we can either merge this PR with a couple of additional tests, or if you want to try to change wf-checking to handle this, we can leave it open. `const_trait_impl` is already marked as an incomplete feature, so merging this PR is absolutely fine.

Reply:
Yeah let's land this first. I'm not sure how the wf scheme would work, but maybe I'll look into that later.